lib/cluster/node.js

```diff
@@ -0,0 +1,979 @@
+     */
+    this.remoteNodes = new Map();
+
+    /**
+     * Cluster nodes activity, used to keep track of nodes being added or
+     * removed, to give more insights to cluster statuses
+     */
+    this.activityMaxLength = this.config.activityDepth;
+    this.activity = [];
+  }
+
+  get syncAddress () {
+    return `tcp://${this.ip}:${this.config.ports.sync}`;
+  }
+
+  async init () {
+    // The publisher needs to be created and initialized before the handshake:
+    // other nodes we'll connect to during the handshake will start to subscribe
+    // to this node right away
+    await this.publisher.init();
+
+    // This also needs to be started before the handshake, as this class handles
+    // direct requests to other nodes (needed to request for the full state
+    // and to introduce oneself to other nodes)
+    await this.command.init();
+
+    global.kuzzle.onPipe('kuzzle:shutdown', () => this.shutdown());
+
+    await this.handshake();
+
+    this.registerEvents();
+    this.registerAskEvents();
+
+    if (this.countActiveNodes() < this.config.minimumNodes) {
+      global.kuzzle.log.info('[CLUSTER] Not enough nodes active. Waiting for other nodes to join the cluster...');
+
+      while (this.countActiveNodes() < this.config.minimumNodes) {
+        await Bluebird.delay(100);
+      }
+    }
+
+    return this.nodeId;
+  }
+
+  /**
+   * Shutdown event: clears all timers, sends a termination status to other
+   * nodes, and removes entries from the cache
+   */
+  async shutdown () {
+    clearInterval(this.heartbeatTimer);
+    await this.idCardHandler.dispose();
+
+    for (const subscriber of this.remoteNodes.values()) {
+      subscriber.dispose();
+    }
+
+    this.publisher.sendNodeShutdown(this.nodeId);
+
+    await this.publisher.dispose();
+    this.command.dispose();
+  }
+
+  /**
+   * Adds a new remote node, and subscribes to it.
+   * @param {string} id            - remote node ID
+   * @param {string} ip            - remote node IP address
+   * @param {number} lastMessageId - remote node last message ID
+   * @return {boolean} false if the node was already known, true otherwise
+   */
+  async addNode (id, ip, lastMessageId) {
+    if (this.remoteNodes.has(id)) {
+      return false;
+    }
+
+    const subscriber = new ClusterSubscriber(this, id, ip);
+
+    this.remoteNodes.set(id, subscriber);
+    await subscriber.init();
+    await subscriber.sync(lastMessageId);
+    await this.idCardHandler.addNode(id);
+
+    global.kuzzle.log.info(`[CLUSTER] Node "${id}" joined the cluster`);
+    this.trackActivity(id, ip, nodeActivityEnum.ADDED);
+
+    if ( global.kuzzle.state === kuzzleStateEnum.NOT_ENOUGH_NODES
+      && this.countActiveNodes() >= this.config.minimumNodes
+    ) {
+      global.kuzzle.state = kuzzleStateEnum.RUNNING;
+      global.kuzzle.log.warn(`[CLUSTER] Minimum number of nodes reached (${this.countActiveNodes()}). This node is now accepting requests again.`);
```

Comment:
```suggestion
      global.kuzzle.log.info(`[CLUSTER] Minimum number of nodes reached (${this.countActiveNodes()}). This node is now accepting requests again.`);
```

Reply:
When the quorum is not reached, a warning is logged about it, telling that Kuzzle won't accept new requests until new nodes join the cluster.
It seems preferable to me that the log line telling that the quorum is reached once again share the same level: if only warnings and above levels of logs are kept, it allows admins to have an insight of when Kuzzle stopped accepting requests, and when it started accepting them again.